### PR TITLE
Update DeviceSpec to work with XCC

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
As part of the migration to the XCC API, the `DeviceSpec` class needs to be updated.

**Description of the Change:**
* The connection object is removed from the `DeviceSpec` class, so that it only contains the specification with which is was initialized. If the spec needs to be refreshed a new `DeviceSpec` object needs to be created.
* The `target` is now retrieved from the device specification rather than passed separately to the `__init__()` method. This avoids any ambiguities.

**Benefits:**
* The `DeviceSpec` class works with XCC.
* The `DeviceSpec` class only encapsulates the device specification (providing some helper methods), and does not handle any connections.

**Possible Drawbacks:**
* The target is now taken from the device specification rather than explicitly passed. Shouldn't cause any issues, but would change the primary target if they had differed. 

**Related GitHub Issues:**
None

### CODE REVIEW

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:34]: I think it would be more accurate to state
```suggestion
            - layout (str): string containing the Blackbird circuit layout or XIR manifest
```

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:51]: I love the idea of the `if` condition here but it turns out to be excessive if we want to avoid duplication:
```suggestion
        missing_keys = {"target", "layout", "modes", "compiler", "gate_parameters"} - spec.keys()
        if missing_keys:
            raise ValueError(
                f"Device specification is missing the following keys: {sorted(missing_keys)}"
            )
```

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:58]: Finally! 🥳

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:71]: I noticed in some of the RemoteEngine tests that `DeviceSpec.compiler` can also just be a string. Is that true?

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:119]: Just to confirm, these are no longer needed, right?

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:107]: Can we add a `Raises` section to this docstring? Otherwise, it's not clear how errors are communicated.

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [inline strawberryfields/api/devicespec.py:125]: Is this right?
```suggestion
    def create_program(self, **parameters: complex) -> sf.Program:
```

File: tests/api/test_devicespec.py
Comment (Mandrenkov) [inline tests/api/test_devicespec.py:25]: What's the difference between `# pylint` and `# pylints`?

File: tests/api/test_devicespec.py
Comment (Mandrenkov) [inline tests/api/test_devicespec.py:153]: Nice! 💯

File: tests/api/test_remote_engine.py
Comment (Mandrenkov) [inline tests/api/test_remote_engine.py:243]: As a heads up, we might not need this after you merge `integrate-xanadu-cloud-client`.

File: tests/frontend/test_program.py
Comment (Mandrenkov) [inline tests/frontend/test_program.py:271]: Are you sure this is legal?
```suggestion
            "compiler": [],
```

**Review by Mandrenkov — COMMENTED** [review]
Looks good to me! 👍

I'll approve once this PR incorporates the changes from #645.

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:51]: Completely right! I added the `missing_keys` later, so I missed this.

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:71]: `_spec["compiler"]` should always be a list of compiler as far as I'm aware. `_spec["default_compiler"]` is a single compiler (the first in the list). Could you point to anywhere where `_spec["compiler"]` is a string?

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:119]: Shouldn't be. They were only used when the TDM spec was returning an unformatted layout, which is no longer the case. These functions aren't called anywhere in neither SF nor OG.

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:125]: Yes! I must confess that I was almost certain that I added this specific return type hint, but apparently I didn't. 😕

File: tests/api/test_devicespec.py
Comment (thisac) [reply tests/api/test_devicespec.py:25]: Oh, whoops, I just added the `s` to check whether these actually needed to be disabled.

File: tests/frontend/test_program.py
Comment (thisac) [reply tests/frontend/test_program.py:271]: Hmm, you're right. Not really, but it made the tests pass (since prior the compiler was set to `None` in the tests).

Comment (thisac) [reply tests/frontend/test_program.py]: Fixed so that `"layout"` is always a string, `"gate_parameters"` is always a dictionary, and `"compiler"` is always a list of strings or empty list.

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [reply strawberryfields/api/devicespec.py:66]: Sure! The tests I was referring to are:
* [`TestRemoteEngineIntegration.test_recompilation_run_async`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L261)
* [`TestRemoteEngineIntegration.test_recompilation_run`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L280)
* [`TestRemoteEngineIntegration.test_validation`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L303)

File: strawberryfields/api/devicespec.py
Comment (Mandrenkov) [reply strawberryfields/api/devicespec.py:123]: Will using a string resolve the circular import issue? 🤔
```suggestion
    def create_program(self, **parameters: complex) -> "sf.Program":
```

**Review by Mandrenkov — APPROVED** [review]
Awesome! 🚀

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:123]: Yes, that could work. But `import strawberryfields as sf` still needs to be added, but never used (which I find a bit ugly just for the sake of a type hint). What do you think?

File: strawberryfields/api/devicespec.py
Comment (thisac) [reply strawberryfields/api/devicespec.py:66]: Just updated them to be lists instead. The tests pass either way, and they should be lists as far as I'm concerned. 🙂

**Pull request merged**
